api/py/ai/chronon/repo/compile.py

```diff
@@ -286,6 +288,58 @@ def _print_features(obj: Union[Join, GroupBy], obj_class: Type[Union[Join, Group
+def _handle_deprecation_warning(
+    obj: Union[Join, GroupBy], obj_class: Type[Union[Join, GroupBy]], downstream_group_bys: dict, downstream_joins: dict
+) -> None:
+    if obj.metaData.deprecationDate:
+        downstream_group_by_with_incorrect_deprecation_date = []
+        for gb_name, gb in downstream_group_bys.items():
+            if not gb.metaData.deprecationDate or gb.metaData.deprecationDate > obj.metaData.deprecationDate:
+                downstream_group_by_with_incorrect_deprecation_date.append(gb_name)
+        if downstream_group_by_with_incorrect_deprecation_date:
+            _print_error(
+                "Deprecation Warning:",
+                f"Downstream group_bys {downstream_group_by_with_incorrect_deprecation_date}"
+                f" are missing deprecationDate or has a later deprecationDate than upstream {obj.metaData.name}."
+                f" Please ensure either to remove/migrate the dependency or set up correct deprecationDate.",
+            )
+        downstream_join_with_incorrect_deprecation_date = []
+        for join_name, join in downstream_joins.items():
+            if not join.metaData.deprecationDate or join.metaData.deprecationDate > obj.metaData.deprecationDate:
+                downstream_join_with_incorrect_deprecation_date.append(join_name)
+            if downstream_join_with_incorrect_deprecation_date:
+                _print_error(
+                    "Deprecation Warning:",
+                    f"Downstream joins {downstream_join_with_incorrect_deprecation_date}"
+                    f" are missing deprecationDate or has a later deprecationDate than upstream"
+                    f" {obj.metaData.name}. Please ensure either to remove/migrate the dependency"
+                    f" or set up correct deprecationDate.",
+                )
+    else:
+        if obj_class is Join:
+            for jp in obj.joinParts:
+                if jp.groupBy.metaData.deprecationDate:
+                    _print_error(
+                        "Deprecation Warning:",
+                        f"Join part {jp.groupBy.metaData.name} is going to be deprecated by"
+                        f" {jp.groupBy.metaData.deprecationDate}. Please ensure either to remove/migrate"
+                        f" the dependency or set up correct deprecationDate for {obj.metaData.name}.",
+                    )
+        else:
+            for source in obj.sources:
+                if source.joinSource and source.joinSource.join.metaData.deprecationDate:
+                    _print_error(
+                        "Deprecation Warning:",
+                        f"Join source {source.joinSource.join.metaData.name} is going to be"
+                        f" deprecated by {source.joinSource.join.metaData.deprecationDate}. Please"
+                        f" ensure either to remove/migrate the dependency or set up correct deprecationDate"
+                        f" for {obj.metaData.name}.",
+                    )
+
```

Comment:
I used GPT to reorganize the code a bit: 🤣 This logic is hard to get right. 
```
def check_deprecation_date(collection, message):
    incorrect_deprecation_date = [
        name for name, item in collection.items()
        if not item.metaData.deprecationDate or item.metaData.deprecationDate > obj.metaData.deprecationDate
    ]

    if incorrect_deprecation_date:
        _print_error(
            "Deprecation Warning:",
            f"Downstream {message} {incorrect_deprecation_date}"
            f" are missing deprecationDate or has a later deprecationDate than upstream {obj.metaData.name}."
            f" Please ensure either to remove/migrate the dependency or set up correct deprecationDate."
        )

def check_deprecation_existence(collection, message):
    for name, item in collection.items():
        if item.metaData.deprecationDate:
            _print_error(
                "Deprecation Warning:",
                f"{message} {name} is going to be deprecated by"
                f" {item.metaData.deprecationDate}. Please ensure either to remove/migrate"
                f" the dependency or set up correct deprecationDate for {obj.metaData.name}.",
            )

if obj.metaData.deprecationDate:
    check_deprecation_date(downstream_group_bys, "group_bys")
    check_deprecation_date(downstream_joins, "joins")
else:
    if obj_class is Join:
        check_deprecation_existence({jp.groupBy.metaData.name: jp.groupBy for jp in obj.joinParts}, "Join part")
    else:
        check_deprecation_existence(
            {source.joinSource.join.metaData.name: source.joinSource.join 
            for source in obj.sources 
            if source.joinSource}, 
            "Join source"
        )
```

Reply:
Updated